[김창곤][뷰:메인페이지, 상세페이지]created main page and info views

xanthangum1 opened the pull request.

Commit:
created main page and info views

xanthangum1 committed:
Merge branch 'main' into feature/movies_info

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+from django.http import JsonResponse
+from django.views import View
+#Custom
+from movie.models import *
```
`import *`을 하게 되면 해당 모듈 내에 있는 모든 객체를 필요하지 않아도 가져오게 됩니다. 따라서, 꼭 필요한 클래스만 지칭하여 import 해주시는 것이 좋습니다.

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+from movie.models import *
+
+
+class ListView(View):
```
창곤님 리스트 컴프리헨션의 형태로 데이터 표출 너무 잘 해주셨고, 객체의 속성으로도 접근 너무 잘 해주셨네요!
그런데 지금 이 영화를 다루는 부분이 whtcha뷰, netflix뷰, korean-best뷰, us-best 뷰 코드가 거의 반복이죠? 거의 대부분 같고 필터링 하는 조건만 다릅니다 :) 서로 다른 옵션을 변수로 받되, 메인 페이지에서는 각 조건에 따라 해당 Movie 표출 함수를 프론트에서 여러번 호출 하는 과정이 진행됩니다. 하나의 코드로 쭉 한 번 합쳐보세요 :)

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+class PageView(View) :
+	def post(self, request) :
+		try:
+			data 	 	= json.loads(request.body)
+			movie_id 	= data['id']
```
- 영화 정보 페이지에 나타날 데이터를 보내주는 엔드포인트죠? 이 함수의 기능은 영화 정보를 받아와서 수정하거나 새로 생성하는 것이 아니라, 영화의 정보를 표출해주는 것입니다. 따라서 `post` 메소드 보다는 `get` 메소드가 적절합니다.

- `get` 메소드를 사용하실 경우 특정 조건 (영화의 id)을 전달 받을 때에는 request의 body가 아닌 query string으로 전달 받습니다

Commit:
Edited pageview to get

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+		try:
+			data 	 	= json.loads(request.body)
+			movie_id 	= data['id']
+			movie_info 	= Movies.objects.get(id = movie_id)
+			photos 		= MoviePhotos.objects.filter(id = movie_id)
+			videos 		= MovieVideos.objects.filter(id = movie_id)
+			cast_list 	= Cast.objects.filter(movie = movie_id)
```
movie와 연결되어 있는 데이터들은 이미 movie 객체의 속성이기 때문에 해당 데이터에 접근 하실 때에 각 클래스를 개별로 사용하지 않습니다. 예를 들어 movie의 id는 `movie.id` 이죠? movie의 타이틀은 `movies.title` 이구요. 그렇다면 무비의 캐스팅 리스트는 어떻게 접근할 수 있을까요?

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+			photos 		= MoviePhotos.objects.filter(id = movie_id)
+			videos 		= MovieVideos.objects.filter(id = movie_id)
+			cast_list 	= Cast.objects.filter(movie = movie_id)
+			actor_id 	= cast_list.all().values_list('id', flat=True)
+			actor_name 	= [People.objects.get(id =person).name for person in cast_list.all().values_list('name', flat=True)]
+			actor_photo	= [People.objects.get(id =person).avatar_url for person in cast_list.all().values_list('name', flat=True)]
```
cast_list가 이미 people 객체가 모여있는 query set 아닌가요? 따라서 person이란 변수에는 개별 people 객체가 담겨 있습니다. 그 객체의 이름을 꺼내오고 싶을 때는 `person` 객체가 가지고 있는 속성을 활용하시면 됩니다. 다시 People 클래스로 가서 get을 하는 것은 이미 사람 객체가 있는데 이름으로 사람 객체를 또 검색하는 것이므로, 같은 동작을 두 번 하는 것입니다 :)

soheon-lee replied:
아래 124, 125번 라인과 같은 방법으로 진행해주세요

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+			videos 		= MovieVideos.objects.filter(id = movie_id)
+			cast_list 	= Cast.objects.filter(movie = movie_id)
+			actor_id 	= cast_list.all().values_list('id', flat=True)
+			actor_name 	= [People.objects.get(id =person).name for person in cast_list.all().values_list('name', flat=True)]
```
여기 동작하나요?

xanthangum1 replied:
네 됩니다!

soheon-lee replied:
`all()` 메소드는 `values_list()` 메소드와 함께 사용하지 않으셔도 됩니다.

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+			actor_id 	= cast_list.all().values_list('id', flat=True)
+			actor_name 	= [People.objects.get(id =person).name for person in cast_list.all().values_list('name', flat=True)]
+			actor_photo	= [People.objects.get(id =person).avatar_url for person in cast_list.all().values_list('name', flat=True)]
+			role_type 	= [person.role for person in cast_list.all()]
+			cast_as 	= [person.cast_as for person in cast_list.all()]
```
good !

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,144 @@
+			actor_photo	= [People.objects.get(id =person).avatar_url for person in cast_list.all().values_list('name', flat=True)]
+			role_type 	= [person.role for person in cast_list.all()]
+			cast_as 	= [person.cast_as for person in cast_list.all()]
+			cast_bank	= [{'name': actor_name[i], 'role': role_type[i], 'cast_as': cast_as[i], 'photo': actor_photo[i]} for i in range(1,len(actor_name))]
```
actor_name을 대상으로 for loop을 돌리면 도는 변수 자체가 이름이기 때문에 `i` 를 사용하지 않을 방법이 있습니다.

Commit:
merged to fix non-fast-forward
Merge branch 'feature/movies_info' of https://github.com/wecode-bootcamp-korea/13-watchandchill-backend into feature/movies_info

Commit:
added desc models.Movies

Commit:
Edited movie.views--PageView

Commit:
views functionified

Commit:
save

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,91 @@
+		library = Movies.objects.all()
+#IN THEATERS
+		library_listed 	= [titles for titles in library]
+		filtered 	   	= [title for title in library if title.in_theaters== True]
```
파이썬에서는 if 조건문 평가시, 반환 되는 값 자체가 boolean 일 경우나, Truthy, Falsy 한 값일 경우,  == False == None 등으로 평가하지 않고, 그 값을 그대로 사용합니다. 그 자체로도 True or False를 판단할 수 있기 때문입니다.

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,91 @@
+			} for movie in filtered_list[:20]]
+			return info
+		library = Movies.objects.all()
+#IN THEATERS
+		library_listed 	= [titles for titles in library]
+		filtered 	   	= [title for title in library if title.in_theaters== True]
+		theater_movies 	= info_lister(filtered)
+#WATCHA ROW
+		filtered 	  	= library.filter(service = '2')
+		watcha_movies 	= info_lister(filtered)
+#NETFLIX ROW
+		filtered		= library.filter(service = '1')
+		netlfix_movies  = info_lister(filtered)
+#KOREA-BEST ROW
+		library_listed 	= [titles for titles in library]
+		filtered 	   	= [title for title in library if title.country == '한국']
+		korean_movies 	= info_lister(filtered)
+#US-BEST ROW
+		library_listed 	= [titles for titles in library]
+		filtered 	   	= [title for title in library if title.country == '미국']
+		us_movies 		= info_lister(filtered)
+#ACTION ROW
+		filtered	= library.filter(genre = '2')
+		action_movies 	= info_lister(filtered)
+#COMEDY ROW
+		filtered 	= library.filter(genre = '12')
+		comedy_movies 	= info_lister(filtered)
```
이 부분 바로 아래 JsonResponse 내부로 들여보내도 좋을 것 같습니다. 불필요한 추가 변수들을 선언하지 않아도 되고, 한 눈에 어떤 키 값에 할당될 지 알 수 있기 때문입니다.

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,91 @@
+		try:
+			data 	 	= json.loads(request.body)
+			movie 	 	= Movies.objects.get(id = movie_id)
+			photos 		= MoviePhotos.objects.filter(id = movie_id)
+			videos 		= MovieVideos.objects.filter(id = movie_id)
+			cast_list 	= Cast.objects.filter(movie = movie_id)
+			actor_name 	= [People.objects.get(id =person).name for person in cast_list.values_list('name', flat=True)]
```
이 엔드포인트 호출 할 때 쿼리문 몇 개 생성되는 지 확인해보셨나요?
이미 65번째 줄에서 movie 객체를 특정했기 때문에, 연결되어 있는 다른 데이터를 가지고 올 때는 그 클래스에 새로 접근 하시지 않고 movie.속성.속성과 같은 단계로 진행하실 수 있습니다. :)

Commit:
edited views

Commit:
views movies

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,120 @@
+			return info
+		library = Movies.objects.all()
+#IN THEATERS
+		library_listed 		= [titles for titles in library]
```
창곤님, 이 리스트 컴프리헨션을 작성하신 의도가 무엇인가요?
정확하게 같은 리스트를 중복해서 또 만들고 있습니다.

soheon-lee commented on movie/views.py:
```diff
@@ -0,0 +1,120 @@
+			return info
+		library = Movies.objects.all()
+#IN THEATERS
+		library_listed 		= [titles for titles in library]
+		theater_filtered 	= [title for title in library if title.in_theaters== 1]
+#WATCHA ROW
+		watcha_filtered 	= library.filter(service = '2')
+#NETFLIX ROW
+		netflix_filtered	= library.filter(service = '1')
+#KOREA-BEST ROW
+		library_listed 		= [titles for titles in library]
+		korea_filtered 		= [title for title in library if title.country == '한국']
+#US-BEST ROW
+		library_listed 		= [titles for titles in library]
+		us_filtered 	   	= [title for title in library if title.country == '미국']
```
이 부분 아래 response 안에 한 번에 다 넣으셔도 좋습니다

Commit:
edited views

Commit:
finished related movie view

Commit:
merged movies to current

Commit:
status progress save

Commit:
Merge branch 'main' into feature/movies_info

Commit:
recovery

Commit:
migrations edit

Commit:
fixed

lordmyshepherd commented on movie/views.py:
```diff
@@ -0,0 +1,113 @@
+class ActorView(View) :
+	def get(self, request, person_id) :
+		try:
+			person 		= People.objects.get(id = person_id)
+			movie_list 	= [{
+				'movie_id'	: cast.movie.id,
+				'title'		: cast.movie.title,
+				'poster_url': cast.movie.poster_url,
+				'services'	: [{'name': info.name, 'logo': info.icon_url} for info in cast.movie.service.all()]
+				} for cast in Cast.objects.filter(name = person)
+			]
+			person_info = {
+				'id'		: person.id,
+				'name'		: person.name,
+				'pic_url' 	: person.avatar_url,
+				'movie_list': list(movie_list)
+			}
+			return JsonResponse({'filmography': person_info}, status=200)
```
select_related()를 사용해서 database에 날리는 query를 줄여보세요~
```
casts = Cast.objects.select_related('movie', 'movie__service')
movies = [{
    'movie_id' : cast.movie.id,
    'title' : cast.movie.id,
    'movie_id' : cast.movie.poster_url,
    'services'	: [{'name': info.name, 'logo': info.icon_url} for info in cast.movie.service.all()]
    } for cast in casts ] 
```

lordmyshepherd commented:
창곤님 view 작성 잘 해주셨어요~~
최고 👍👍👍👍

comment 남긴 내용은 프로젝트 끝나고 수정해보세요~
